core/lib/zksync_core/src/state_keeper/io/seal_logic/mod.rs

```diff
@@ -269,10 +269,16 @@ impl UpdatesManager {
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum SealStrategy<'a, 'b> {
```

Comment:
Nit: To make dealing with this type simpler, you can take `Connection` by value; in the only place when the `Sequential` variant is constructed, you can obtain the connection later by destructuring the enum. This has a downside that it's possible to forget to commit the transactional connection at the end, but since it's used in a single place, IMO, it's manageable.

Here's another optimization that would simplify `SealStrategy` use. Since it's not used that extensively, feel free to ignore.

```rust
#[derive(Debug)]
pub(crate) enum SealStrategy<'pool> {
    Sequential(Connection<'pool, Core>),
    Parallel(&'pool ConnectionPool<Core>),
}

// As opposed to `Cow` from `std`; a union of an owned type and a mutable ref to it
enum Goat<'a, T> {
    Owned(T),
    Borrowed(&'a mut T),
}

impl<T> ops::Deref for Goat<'_, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        match self {
            Self::Owned(value) => value,
            Self::Borrowed(value) => value,
        }
    }
}

impl<T> ops::DerefMut for Goat<'_, T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        match self {
            Self::Owned(value) => value,
            Self::Borrowed(value) => value,
        }
    }
}

impl<'pool> SealStrategy<'pool> {
    async fn connection(&mut self) -> anyhow::Result<Goat<'_, Connection<'pool, Core>>> {
        Ok(match self {
            Self::Parallel(pool) => {
                Goat::Owned(pool.connection_tagged("state_keeper").await?)
            }
            Self::Sequential(conn) => Goat::Borrowed(conn),
        })
    }
}

// Then, you can simplify strategy uses like this:
if self.pre_insert_txs {
    let progress = L2_BLOCK_METRICS.start(L2BlockSealStage::PreInsertTxs, is_fictive);
    let mut connection = strategy.connection().await?;
    self.insert_transactions(&mut connection)
        .await
        .context("failed persisting transactions in L2 block")?;
    progress.observe(Some(self.l2_block.executed_transactions.len()));
}
```

The trick here is to have a meaningful lifetime in `SealStrategy` (namely, that of a connection pool); otherwise, it would be issues coecring `&mut Connection<'a, Core>` to `&mut Connection<'b, Core>` with `'a != 'b` because mutable references are invariant w.r.t. the reference type (which includes `'a` / `'b` in this case).